src/test/java/org/tarantool/ClientReconnectClusterIT.java

```diff
@@ -393,6 +395,53 @@ void testDelayFunctionResultFetch() {
         expectDisconnected(client, spaceId, pkId);
     }
 
+    @Test
+    void testRoundRobinSocketProviderRefused() {
+        stopInstancesAndAwait(SRV1);
+        stopInstancesAndAwait(SRV2);
+        stopInstancesAndAwait(SRV3);
+
+        RuntimeException error = new RuntimeException("Fake error");
+        TarantoolClusterClientConfig config = makeDefaultClusterClientConfig();
+        config.initTimeoutMillis = 1000;
+        Throwable exception = assertThrows(
+            CommunicationException.class,
+            () -> {
+                new TarantoolClusterClient(
+                    config,
+                    TestUtils.wrapByErroredProvider(new RoundRobinSocketProviderImpl(
+                        "localhost:" + PORTS[0],
+                        "localhost:" + PORTS[1],
+                        "localhost:" + PORTS[2]
+                    ), error)
+                );
+            }
+        );
+        assertTrue(findCause(exception, error));
+    }
+
+    @Test
+    void testRoundRobinSocketProviderRefusedAfterConnect() {
+        final TarantoolClientImpl client = makeClusterClient(
+            "localhost:" + PORTS[0],
+            "localhost:" + PORTS[1],
+            "localhost:" + PORTS[2]
+        );
+
+        client.ping();
+        stopInstancesAndAwait(SRV1);
+
+        client.ping();
+        stopInstancesAndAwait(SRV2);
+
+        client.ping();
+        stopInstancesAndAwait(SRV3);
+
+        CommunicationException exception = assertThrows(CommunicationException.class, client::ping);
+        Throwable origin = exception.getCause();
+        assertEquals(origin, client.getThumbstone());
```

Comment:
I would also check that the tombstone contains ConnectException in its causes.

Reply:
There is no ConnectException. Moreover, the real cause depends on which operation fails first (reconnection, reading, writing). In my case it is always `CommunicationException: Cant read answer` due to the reading thread tries receiving a response before other threads do anything.

Reply:
How can we try to read something if there is no connected socket to read from?

Reply:
For instance,
1. client has an established connection to an instance.
2. reading thread blocks on reading operation.
3. the instance is down.
4. reading thread is interrupted by exception that socket has died.
5. client fills thumbstone with the error.

Reply:
Okay, it at least doesn't seems to be easy to ensure a certain error in the case (and don't sure we should). So it looks as the thing to consider in #166.

However when we connect a first time we definitely will receive ConnectException. Are not you think it worth to check?

BTW, I looked into the stack trace from this test; it looks so:

```
org.tarantool.CommunicationException: Cant read answer
	at org.tarantool.TarantoolClientImpl.die(TarantoolClientImpl.java:299)
	at org.tarantool.TarantoolClientImpl.readThread(TarantoolClientImpl.java:422)
	at org.tarantool.TarantoolClientImpl.lambda$startThreads$0(TarantoolClientImpl.java:200)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.tarantool.CommunicationException: Channel read failed: -1
	at org.tarantool.protocol.ReadableViaSelectorChannel.read(ReadableViaSelectorChannel.java:47)
	at org.tarantool.protocol.ProtoUtils.readPacket(ProtoUtils.java:78)
	at org.tarantool.TarantoolClientImpl.readThread(TarantoolClientImpl.java:412)
	... 2 more
```

And it seems that we can give a user more clear error here and say EOF instead of -1.